### ISSUE TITLE
`push` default behavior is to push all the data -- IMHO undesirable in many cases and unsafe

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#4551]
`publish` by default copied no data to the remote unless specific files were pointed to.

`push` seems to try to publish all the data right away.  

<details>
<summary>Script adjusted from the one posted in https://github.com/datalad/datalad/issues/4550#issuecomment-630268894 but without wanted set</summary> 

```shell
$> cat ./check-push-content-nowanted.sh
#!/bin/bash

export PS4='> '
set -eu
cd "$(mktemp -d ${TMPDIR:-/tmp}/dl-XXXXXXX)"
set -x

datalad create src

(
cd src
# Produce and annotate some data
echo -e '\x00' > data.0
echo -e '\x01' > secure.1
echo -e '\x02' > secure.2
datalad save -m "added some files"

# Dropping a file to mimic a case of simply not having it locally (thus not to be "pushed")
git annex drop --force secure.2

# Annotate sensitive content, actual value "verysecure" does not matter in this example
git annex metadata --set distribution-restrictions=verysecure secure.1 secure.2

datalad save -m "added a file"
datalad create-sibling -r -s target ../target
datalad push --to target
)

(
cd target
git annex list
)
```
</details>

which results in
```shell
> datalad push --to target
publish(ok): . (dataset) [refs/heads/master->target:refs/heads/master [new branch]]                            
[WARNING] Slated for transport, but no content present [copy(/home/yoh/.tmp/dl-npppKEt/src/secure.2)] 
copy(impossible): secure.2 (symlink) [Slated for transport, but no content present]
copy(ok): data.0 (file) [to target...]                                                                         
copy(ok): secure.1 (file) [to target...]
publish(ok): . (dataset) [refs/heads/git-annex->target:refs/heads/git-annex 369c340..a4ffe63]                
```

I see that in some cases such default behavior would be desired, and warnings about some files be impossible to transfer are due. But that is IMHO an undesirable default behavior:

- quite often people do not have all the files in the repository (containers etc), and want to publish what they have -- complaining about files they do not have be just wrong and distracting (could be many more from what they actually desire to share/publish)
- I might not want to publish stuff which is already available elsewhere (e.g. containers)
- only a matter of time until pushing of all data by default would leak some sensitive data into public space 
- I typically do not push all the data by default. I rely on `wanted` set and `--auto` (see further discussion on --auto in #4550) 

Such default behavior is a radical change from `publish` behavior where by default no data was published, so I could feel save running `datalad publish -r` on a hierarchy.  With "copy by default" - I would not dare ("did I forget to set restrictive 'wanted' to some repo?  did I set wanted good enough to not transfer some containers I would prefer to not copy but do have locally?")

```shell
$> datalad --version
datalad 0.13.0rc1.dev52
```

edit: fixed referenced issue to be #4550 not 4541

**Comment by kyleam** [comment]
> * quite often people do not have all the files in the repository (containers etc), and want to publish what they have -- complaining about files they do not have be just wrong and distracting (could be many more from what they actually desire to share/publish)

This is related to gh-4508.  gh-4547 restricts the warning to when a path is explicitly given.  Does that address your concern?

**Comment by yarikoptic** [comment]
I could have said yes - for my immediate use case where I provide no paths. But my concern is more generic: issuing a warning on a (sub)path for which I have no content. `push` ATM just says that providing a path restricts operation to that. It is a clear procedural description. Warning would be in place for the cases where we also provide guarantee that everything for that path is present in remote after operation. It is orthogonal to restricting operation to subtree, and afaik we didn't discuss how we let user to establish such guarantees where desired.

**Comment by yarikoptic** [comment]
per @mih request: here is code in publish which decides on either to add `--auto`:
https://github.com/datalad/datalad/blob/master/datalad/distribution/publish.py#L190

**Comment by yarikoptic** [comment]
The decision was to continue with such default behavior of pushing the data, no issue is to "resolve" this issue's main point so closing.  
Workaround to be implemented in #4581 for people who setup their repository with `wanted` settings is probably just to set `datalad.push.copy-auto-if-wanted` in global configuration (if they RTFM carefully) to prevent any  undesired data copying by accident of using `datalad push` instead of e.g. `datalad publish`.
